### ISSUE TITLE
BUG: scl_slope/scl_inter in mask and dseg headers are set to nan, should be 1/0

### ISSUE THREAD

**Issue by lorisnaspi** [nipreps/fmriprep#2507]
Dear fmriprep experts,

I have just finished the preprocessing and everything seems alright since no errors were reported.
However, I have noticed that I cannot open the “sub-XX_desc-brain_mask.nii” file.

I use MRIcron to open these images and when I click on it it says “Invalid floating point operation. Press OK to ignore and risk data corruption. Press Abort to kill the program”.

Basically, I cannot even explore the mask since it contains NaN. Any idea why that is the case? 
I have notices WARNINGS during the anatomical and functional processing in which data types were converted from float32 to uint8 for the brain masks. These are the only files I cannot open. I wonder whether the two things are related

I am looking forward to hearing from your opinions
best wishes
Loris

**Comment by effigies** [comment]
Would you be able to share the mask?

**Comment by lorisnaspi** [comment]
Hi,

These are two of the files in the "anat" folder that won't open. They are
attached below.
Please let me know if you have any idea about what went wrong.

Loris

**Comment by effigies** [comment]
GitHub strips attachments. Please upload via the web interface.

**Comment by lorisnaspi** [comment]
Hi
Apologies for that. I will attach them through the web. I hope you can see them now.
[sub-3000_desc-brain_mask.nii.gz](https://github.com/nipreps/fmriprep/files/7015771/sub-3000_desc-brain_mask.nii.gz)
[sub-3000_space-MNI152NLin2009cAsym_desc-brain_mask.nii.gz](https://github.com/nipreps/fmriprep/files/7015775/sub-3000_space-MNI152NLin2009cAsym_desc-brain_mask.nii.gz)
[sub-3000_space-MNI152NLin2009cAsym_dseg.nii.gz](https://github.com/nipreps/fmriprep/files/7015778/sub-3000_space-MNI152NLin2009cAsym_dseg.nii.gz)

Can you open them now?
Loris

**Comment by effigies** [comment]
I also see these fail to open with mricron, but they look reasonable in nibabel:

```Python
In [1]: import nibabel as nb

In [2]: import numpy as np

In [3]: mask1 = nb.load('sub-3000_desc-brain_mask.nii.gz')

In [4]: mask2 = nb.load('sub-3000_space-MNI152NLin2009cAsym_desc-brain_mask.nii.
   ...: gz')

In [5]: dseg = nb.load('sub-3000_space-MNI152NLin2009cAsym_dseg.nii.gz')

In [6]: np.unique(mask1.dataobj)
Out[6]: array([0, 1], dtype=uint8)

In [7]: np.unique(mask2.dataobj)
Out[7]: array([0, 1], dtype=uint8)

In [8]: np.unique(dseg.dataobj)
Out[8]: array([0, 1, 2, 3], dtype=int16)
```

And using `mask1.orthoview()`, etc...

![mask1](https://user-images.githubusercontent.com/83442/130102381-022cfd36-56a2-4aaa-829e-6efee95b45b1.png)
![mask2](https://user-images.githubusercontent.com/83442/130102382-38ed73a0-c1c7-4cd3-87c5-6cba023289f2.png)
![dseg](https://user-images.githubusercontent.com/83442/130102380-9911fd55-b5ec-49e1-9e7a-3028d4a985e2.png)

So this seems to be a bug in mricron.

**Comment by lorisnaspi** [comment]
Hi,

Thank you very much for this. This is a relief somehow because it means the entire analysis was OK.
Can I ask you if nibabel has a gui similar to MRIcron to open these images? I mean I come from Matlab and R and I have just strated Python.  It would simplify my life a lot. Or alternative solutions?

Thank you very much
Loris

**Comment by effigies** [comment]
You can view an image using nibabel like so:

```Python
import nibabel as nb
from matplotlib import pyplot as plt

img = nb.load("/path/to/file")
img.orthoview()

plt.show()  # This is typically required, but sometimes img.orthoview() will pop up a window without making this call
```

**Comment by lorisnaspi** [comment]
Great. Thank you. Shall we close the thread?
Loris.

**Comment by effigies** [comment]
Sounds good. Would you mind posting to https://github.com/neurolabusc/MRIcron/issues as well?

**Comment by neurolabusc** [comment]
MRIcron is showing these images accurately. The plots shown by @effigies do not take account for the [scl_slope and scl_inter](https://nifti.nimh.nih.gov/pub/dist/src/niftilib/nifti1.h).

Any value multiplied by or added to a Not-A-Number is necessarily not a number. You can see the issue by examining the header: 

```
fslhd sub-3000_desc-brain_mask.nii
...
scl_slope	nan
scl_inter	nan
```

The NIfTI file is explicitly stating that all voxels are not a number. I suspect this is a limitation of the images, and the intended scl_slope = 1.0 and scl_inter = 0. In this case, the images are incorrect. MRIcron is scaling the images as required by the standard.

**Comment by neurolabusc** [comment]
SPM also detects these images are defective:

```
>> hdr = spm_vol( 'sub-3000_desc-brain_mask.nii')
Error using  ~ 
NaN's cannot be converted to logicals.

Error in nifti (line 63)
        if ~vol.hdr.scl_slope && ~vol.hdr.scl_inter

Error in spm_vol_nifti (line 19)
    N = nifti(P);

Error in spm_vol>spm_vol_hdr (line 128)
V = spm_vol_nifti(p,n);

Error in spm_vol (line 61)
        v = spm_vol_hdr(deblank(P(i,:)));
```

**Comment by lorisnaspi** [comment]
> MRIcron is showing these images accurately. The plots shown by @effigies do not take account for the [scl_slope and scl_inter](https://nifti.nimh.nih.gov/pub/dist/src/niftilib/nifti1.h).
> 
> Any value multiplied by or added to a Not-A-Number is necessarily not a number. You can see the issue by examining the header:
> 
> ```
> fslhd sub-3000_desc-brain_mask.nii
> ...
> scl_slope	nan
> scl_inter	nan
> ```
> 
> The NIfTI file is explicitly stating that all voxels are not a number. I suspect this is a limitation of the images, and the intended scl_slope = 1.0 and scl_inter = 0. In this case, the images are incorrect. MRIcron is scaling the images as required by the standard.

Thank you for this.
That is what I also tried with SPM. I knew the mask contained NaNs but I do not understand why. Is there a way to turn around this problem with MRIcron? I need these masks to go on  with skull stripping and the analysis but if I cannot open them with SPM...it is bad. Any advice?

Thank you
Loris

**Comment by effigies** [comment]
I had checked this, but I saw:

```
$ fslhd sub-3000_desc-brain_mask.nii.gz
filename       sub-3000_desc-brain_mask.nii.gz
[...]
scl_slope      0.000000
scl_inter      0.000000
```

From the standard:

> ```
>    If the scl_slope field is nonzero, then each voxel value in the dataset
>    should be scaled as
>       y = scl_slope * x + scl_inter
>    where x = voxel value stored
>          y = "true" voxel value
> ```

Being a bit more careful:

```Python
In [1]: import nibabel as nb

In [2]: import gzip

In [3]: with gzip.open('sub-3000_desc-brain_mask.nii.gz', 'rb') as fobj:
   ...:     hdr = nb.Nifti1Header.from_fileobj(fobj)

In [4]: print(hdr)
<class 'nibabel.nifti1.Nifti1Header'> object, endian='<'
[...]
scl_slope       : nan
scl_inter       : nan
[...]
```

So it seems we need to ensure that we set slope/inter to 1/0.

In the meantime, you can fix this with:

```Python
import nibabel as nb

img = nb.load(fname, mmap=False)
img.header.set_slope_inter(slope=1, inter=0)
nb.Nifti1Image(np.array(img.dataobj), None, img.header).to_filename(fname)
```

**Comment by lorisnaspi** [comment]
Thank you @effigies

Just to let you aware that the same error is also in the "func" folder for the brain mask, not only in thr "anat"
Also, does that line of code also resave the mask with slope = 1 and inter = 1, i.e. overwrite the previously wrong header?

Thank you for the support
Loris

**Comment by effigies** [comment]
@lorisnaspi Apologies, I didn't see that. The line of code would overwrite the previously wrong header.

This issue will be fixed in 20.2.4 and 21.0.0.

**Comment by eburkevt** [comment]
Has this bug been fixed in version 20.2.4 onwards?

**Comment by effigies** [comment]
Hi, yes. This has been fixed in 20.2.4 and up.
